### PR TITLE
v4: objection v1 and plugin ownership

### PR DESCRIPTION
**Pull request by devinivy**
- [x] Upgrade to objection v1 (#53)
 - [x] Refactor plugin ownership of knex and models
 - [x] Tests
 - [x] Docs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,8 +11,11 @@ A [hapi](https://github.com/hapijs/hapi) plugin integrating [Objection ORM](http
 Schwifty is used to define [Joi](https://github.com/hapijs/joi)-compatible models and knex connections for use with Objection ORM.  Those models then become available within your hapi server where it is most convenient.  It has been tailored to multi-plugin deployments, where each plugin may set clear boundaries in defining its own models, knex database connections, and migrations.  It's safe to register schwifty multiple times, wherever you'd like to use it, as it protects against model name collisions and other ambiguous configurations.
 
 ```js
-// First, ensure your project includes your
-// preferred versions of knex, objection, and sqlite3
+// First, ensure your project includes knex, objection, and sqlite3
+// Note that for knex and Objection, we assume using
+// knex >=0.8 and Objection >=1 <2 (see our peerDependencies in our package file)
+// Schwifty v4 introduced compatibility with Objection v.1
```
Comment (mattboutet) [inline README.md]: It might just be me, but `v.1` reads like `v0.1` to me, not `v1.0` -- might be good to tweak this to disambiguate. Maybe just `v1`?

Comment (devinivy) [reply README.md]: Agree 👍

File: API.md [hunk 2]
```diff
@@ -17,14 +17,18 @@ Schwifty takes the following registration options,
 Returns the knex instance used by the current plugin. In other words, this returns the knex instance for the active realm, falling back to the root server's server-wide default knex instance.
 
 #### `server.models([all])`
-Returns an object containing models keyed by `name`.  When `all` is `true`, models across the entire server are returned.  Otherwise, only models declared within the current plugin (or, active realm) are returned.
+Returns an object containing models keyed by `name`.  When `all` is `true`, models across the entire server are returned.  Otherwise, only models declared within a.) the current plugin (or, server's active realm) and b.) any plugins for which the current plugin is an ancestor (e.g. if the current plugin has registered a plugin that registers models) are returned.
+
+
+The services that are available on this object are only those registered by server or .
```
Comment (devinivy) [inline API.md]: Dangles?

File: API.md [hunk 3]
```diff
@@ -34,7 +38,14 @@ Used to register models, knex instances, and migration directory information on
 Returns the knex instance used by `request.route`'s plugin. In other words, this returns the knex instance for `request.route`'s active realm, falling back to the root server's server-wide default knex instance.
 
 #### `request.models([all])`
-Returns an object containing models keyed by `name`.  When `all` is `true`, models across the entire server are returned.  Otherwise, only models declared within `request.route`'s plugin (or, active realm) are returned.
+Returns an object containing models keyed by `name`.  When `all` is `true`, models across the entire server are returned. Otherwise, only models declared within a.) `request.route`'s plugin (or, active realm) and b.) any plugins for which `request.route`'s plugin is an ancestor (e.g. if `request.route`'s plugin has registered a plugin that registers models) are returned.
+
+### Response toolkit decorations
+#### `h.knex()`
+Returns the knex instance used by the current route's plugin. In other words, this returns the knex instance for the active realm as identified by `h.realm`, falling back to the root server's server-wide default knex instance.
```
Comment (devinivy) [inline API.md]: Let's chat about this and find a good way to describe the new behavior 👍

File: README.md [hunk 4]
```diff
@@ -11,8 +11,11 @@ A [hapi](https://github.com/hapijs/hapi) plugin integrating [Objection ORM](http
...
+// If you prefer / need to work with earlier versions of Objection, checkout Schwifty v3 instead!
 
 // To get started you might run,
 // npm install --save hapi@17 joi schwifty knex objection sqlite3
```
Comment (devinivy) [inline README.md]: We can get rid of `@17` now that it has been fully released.

File: README.md (hunk 1 above)
Comment (zemccartney) [reply README.md]: 👌 mm! dah, good point. thanks, y'all, fixing now 🙏

File: API.md (hunk 2 above)
Comment (zemccartney) [reply API.md]: aegrfkajwefgkewugu3qi extra dangles. sorry, slopslopslop, I'll clear that up

File: README.md (hunk 4 above)
Comment (zemccartney) [reply README.md]: ahh gotcha. cool, thanks for catching that. i'll fix 'er up 🙏

File: API.md (hunk 3 above)
Comment (zemccartney) [reply API.md]: 👌 that'd be great. thanks, my dude. i'll bug ya

**Comment by zemccartney** [comment]
@devinivy yo! took a crack at clearing things up re: how knex instances are bound to models. no idea how I did 🎺 thanks for the help today 🙏

File: lib/index.js [hunk 5]
```diff
@@ -112,25 +104,22 @@ internals.registrationConfig = (options, rootRealm) => {
...
-    collector.knexGroups.forEach((knexGroup) => {
+    collector.realmByModel.forEach((realm, Model) => {
 
-        const models = knexGroup.models;
-        const knex = knexGroup.knex || rootKnex;
+        const knex = internals.knex(() => realm)();
+        const BoundModel = (knex && !Model.knex()) ? Model.bindKnex(knex) : Model;
 
-        models.forEach((modelName) => {
+        Toys.forEachAncestorRealm(realm, (r) => {
```
Comment (wswoodruff) [inline lib/index.js:114]: :heart:

**Comment by crespowang** [comment]
really can't wait for 4.0.0 release! things are broken after upgrading to Objection 1.0.0

**Comment by devinivy** [comment]
@crespowang feel free to use this branch in the short term.  The code is stable and well-tested—we're just working on the docs at this point.  Expect a release in the next day or two!

**Pull request merged**